### PR TITLE
Http exploit for windows Apache version 2.4.49(New Module) Tested

### PR DESCRIPTION
**Pull request by taha-ishaq**
Name: Apache HTTP Server 2.4.49 Path Traversal
Description: This Metasploit module exploits the Path Traversal vulnerability in Apache HTTP Server version 2.4.49 (CVE-2021-41773). The vulnerability allows an attacker to traverse directories and read arbitrary files on the target system by sending crafted HTTP requests to the server. This module automates the exploitation process by taking a list of target URLs and a traversal path as input, attempting to exploit the vulnerability on each target. If successful, it retrieves the contents of the specified file on the target system.
Author: taha-ishaq
License: MSF_LICENSE
Platform: Unix, Linux, Windows

Metasploit Framework (MSF) Usage:
Load the Module:
Load the module into the Metasploit Framework using the use command.
http1

bash
Copy code
msf6 > use exploit/unix/http/saints_exploit
Set Options:
Set the required options for the exploit. These options include:

TARGETS_FILE: Path to a file containing a list of target URLs.
TRAVERSAL_PATH: Path to the file to traverse on the target system.
bash
Copy code
msf6 exploit(unix/http/saints_exploit) > set TARGETS_FILE /path/to/targets.txt
msf6 exploit(unix/http/saints_exploit) > set TRAVERSAL_PATH /etc/passwd
Run the Exploit:
Execute the exploit by running the run command.

http2(updated)

scss
Copy code
msf6 exploit(unix/http/saints_exploit) > run
Review Results:
Review the output to see if the exploit was successful on any of the target systems. Successful traversal attempts will display the contents of the specified file on the target system.

Module Behavior:
The module takes a list of target URLs and a traversal path as input.
It sends crafted HTTP requests to each target URL, attempting to exploit the Path Traversal vulnerability.
If successful, it retrieves the contents of the specified file on the target system.
The exploit results are displayed in the Metasploit console, indicating successful or failed exploitation attempts.
http3(updated)

http3(updated)

THis exploit and vulnerability is not in the exploit db of metasploit and is also a working one test on real site !
+ I am a student and new to metasploit community , Positive feedbacks and guidance from the best is always welcome . Thanks!

### CODE REVIEW

**Comment by adfoster-r7** [comment]
Thanks for the pull request; I've just searched for the version of the software that you're targeting and it looks like it already exists in framework:

```
msf6 auxiliary(scanner/http/apache_normalize_path) > search Apache HTTP Server version 2.4.49

Matching Modules
================

   #  Name                                          Disclosure Date  Rank    Check  Description
   -  ----                                          ---------------  ----    -----  -----------
   0  auxiliary/scanner/http/apache_normalize_path  2021-05-10       normal  No     Apache 2.4.49/2.4.50 Traversal RCE scanner


Interact with a module by name or index. For example info 0, use 0 or use auxiliary/scanner/http/apache_normalize_path

msf6 auxiliary(scanner/http/apache_normalize_path) > 
```

Are there any differences to this new module and the existing module that framework offers? 👀

**Comment by taha-ishaq** [comment]
Well I checked the modules you provided and it seems there is a bit of difference with them and the one I provided , can you kindly check this yourself please as the results are of the same site (the one i exploited in web bounty) THe specter one is mine and others are the ones that i wasnt able to find before  
fault1updated

fault 3updated

fault4updated
fault5updated
faut2

**Comment by taha-ishaq** [comment]
> Thanks for the pull request; I've just searched for the version of the software that you're targeting and it looks like it already exists in framework:
> 
> ```
> msf6 auxiliary(scanner/http/apache_normalize_path) > search Apache HTTP Server version 2.4.49
> 
> Matching Modules
> ================
> 
>    #  Name                                          Disclosure Date  Rank    Check  Description
>    -  ----                                          ---------------  ----    -----  -----------
>    0  auxiliary/scanner/http/apache_normalize_path  2021-05-10       normal  No     Apache 2.4.49/2.4.50 Traversal RCE scanner
> 
> 
> Interact with a module by name or index. For example info 0, use 0 or use auxiliary/scanner/http/apache_normalize_path
> 
> msf6 auxiliary(scanner/http/apache_normalize_path) > 
> ```
> 
> Are there any differences to this new module and the existing module that framework offers? 👀

yes I commented all that you can check sorry , I accidently commented instead of replying

**Comment by adfoster-r7** [comment]
Sorry I have not had a chance to confirm; I just know that the existing module also supports actions:

```
msf6 auxiliary(scanner/http/apache_normalize_path) > show actions

Auxiliary actions:

       Name             Description
       ----             -----------
       CHECK_RCE        Check for RCE (if mod_cgi is enabled).
   =>  CHECK_TRAVERSAL  Check for vulnerability.
       READ_FILE        Read file on the remote server.

msf6 auxiliary(scanner/http/apache_normalize_path) > set action READ_FILE
action => READ_FILE
```

Does this module add a new action that isn't supported in the existing module?

**Comment by taha-ishaq** [comment]
> Sorry I have not had a chance to confirm; I just know that the existing module also supports actions:
> 
> ```
> msf6 auxiliary(scanner/http/apache_normalize_path) > show actions
> 
> Auxiliary actions:
> 
>        Name             Description
>        ----             -----------
>        CHECK_RCE        Check for RCE (if mod_cgi is enabled).
>    =>  CHECK_TRAVERSAL  Check for vulnerability.
>        READ_FILE        Read file on the remote server.
> 
> msf6 auxiliary(scanner/http/apache_normalize_path) > set action READ_FILE
> action => READ_FILE
> ```
> 
> Does your new module add a new action that isn't supported in the existing module?

Fair enough I guess, Can you guide me for a module to make thats not already in metasploit, I dont want to make something that already exists again, Thanks !

**Comment by taha-ishaq** [comment]
> Sorry I have not had a chance to confirm; I just know that the existing module also supports actions:
> 
> ```
> msf6 auxiliary(scanner/http/apache_normalize_path) > show actions
> 
> Auxiliary actions:
> 
>        Name             Description
>        ----             -----------
>        CHECK_RCE        Check for RCE (if mod_cgi is enabled).
>    =>  CHECK_TRAVERSAL  Check for vulnerability.
>        READ_FILE        Read file on the remote server.
> 
> msf6 auxiliary(scanner/http/apache_normalize_path) > set action READ_FILE
> action => READ_FILE
> ```
> 
> Does your new module add a new action that isn't supported in the existing module?

Well I checked again with the same web and the module you provided still cant exploit the target even with all its actions chosen but mine did , I dont know but there's this vulnerablity where you can rea the target files that give 500 status code using a path traversal technique, The results are same as the ones I provided before , The already existing module still says that the target is not vulnerable but it indeed is and can be exploited using my module, So can you recheck and make your own report on both modules, Thanks!

**Comment by adfoster-r7** [comment]
> The already existing module still says that the target is not vulnerable but it indeed is and can be exploited using my module

If that's the case, would it make sense to update the existing module with a new action compromised of your logic changes? Or maybe it would make sense to find out what the difference is - maybe there's a bug in the existing module

**Comment by taha-ishaq** [comment]
> > The already existing module still says that the target is not vulnerable but it indeed is and can be exploited using my module
> 
> If that's the case, would it make sense to update the existing module with a new action compromised of your logic changes? Or maybe it would make sense to find out what the difference is - maybe there's a bug in the existing module

anything you suggest sir but adding a path traversal module for apache isnt a bad choice either

**Comment by taha-ishaq** [comment]
> > The already existing module still says that the target is not vulnerable but it indeed is and can be exploited using my module
> 
> If that's the case, would it make sense to update the existing module with a new action compromised of your logic changes? Or maybe it would make sense to find out what the difference is - maybe there's a bug in the existing module

Any updates sir

**Comment by taha-ishaq** [comment]
> > The already existing module still says that the target is not vulnerable but it indeed is and can be exploited using my module
> 
> If that's the case, would it make sense to update the existing module with a new action compromised of your logic changes? Or maybe it would make sense to find out what the difference is - maybe there's a bug in the existing module

These two Metasploit modules exploit different vulnerabilities in Apache HTTP Server (CVE-2021-41773 and CVE-2021-42013) and have distinct functionality and usage.

Here's a breakdown of the differences between the two modules:

Vulnerability Exploited:

The first module exploits the Path Traversal vulnerability (CVE-2021-41773) in Apache HTTP Server version 2.4.49.
The second module exploits an unauthenticated Remote Code Execution (RCE) vulnerability in Apache HTTP Server versions 2.4.49 and 2.4.50. This vulnerability has two CVEs associated with it: CVE-2021-41773 and CVE-2021-42013.
Description:

The first module focuses on exploiting the Path Traversal vulnerability to traverse directories and potentially read sensitive files outside of the document root.
The second module targets the RCE vulnerability, allowing execution of arbitrary commands on the target system.
Functionality:

The first module attempts to exploit the Path Traversal vulnerability by crafting specific URIs to trigger directory traversal.
The second module provides options to choose the CVE to exploit, supports depth for path traversal, and offers different payloads for exploitation based on the selected target.
Payloads:

The first module does not generate payloads. It aims to exploit the vulnerability directly by traversing directories.
The second module provides payloads tailored for different purposes, such as dropping a Meterpreter shell on Linux systems or executing arbitrary Unix commands in-memory.
Default Options:

The first module does not provide default options related to payload selection or exploitation method.
The second module sets default options for choosing the vulnerability to exploit, the depth for path traversal, and the target URI.
In summary, while both modules target vulnerabilities in Apache HTTP Server, they exploit different weaknesses and offer distinct functionalities tailored to their respective vulnerabilities. The first module focuses on directory traversal, while the second module provides options for remote code execution 

These are the differences between these 2 exploits so far, So what are your thoughts kindly share

**Comment by cdelafuente-r7** [comment]
Thanks @taha-ishaq for your contribution.

The module [`auxiliary/scanner/http/apache_normalize_path`](https://github.com/rapid7/metasploit-framework/blob/95283f7d3e35ce7080359375c1a914d1ace487c9/modules/auxiliary/scanner/http/apache_normalize_path.rb) is an [auxiliary/scanner](https://docs.metasploit.com/docs/development/developing-modules/guides/how-to-get-started-with-writing-an-auxiliary-module.html) module and the module [`exploit/multi/http/apache_normalize_path_rce`](https://github.com/rapid7/metasploit-framework/blob/95283f7d3e35ce7080359375c1a914d1ace487c9/modules/exploits/multi/http/apache_normalize_path_rce.rb) is an [exploit](https://docs.metasploit.com/docs/development/developing-modules/guides/get-started-writing-an-exploit.html) module. Both target CVE-2021-41773 and CVE-2021-42013 vulnerabilities that affect Apache version 2.4.49. The goal of an auxiliary/scanner module is different from an exploit module, that's why they both are written in two separate modules.

An auxiliary module will leverage the vulnerability to enumerate or retrieve data from the target. An exploit module will execute a payload, and, in this case, get remote code execution.

As you mentioned, both modules have multiple options available that enable a user to choose which CVE to use. The auxiliary module also has multiple `ACTION`'s available. Please refer to their documentations [here](https://github.com/rapid7/metasploit-framework/blob/95283f7d3e35ce7080359375c1a914d1ace487c9/documentation/modules/auxiliary/scanner/http/apache_normalize_path.md) and [here](https://github.com/rapid7/metasploit-framework/blob/95283f7d3e35ce7080359375c1a914d1ace487c9/documentation/modules/exploit/multi/http/apache_normalize_path_rce.md).

Your module looks like it is targeting the same vulnerability (CVE-2021-41773) and retrieve files from the target. This is exactly what the `auxiliary/scanner/http/apache_normalize_path` module with the action `READ_FILE` seems to do, unless I'm missing something.

As it has been said before, if you spot an issue in the existing module or if you think they could be improved by adding new techniques, you're more than welcome to update them. We will help you to do so if you have questions. Also, you can refer to the [documentation](https://docs.metasploit.com/docs/development/developing-modules/) about writing modules.

One last thing, we do not accept two modules in a single PR. Please submit two separate PR's with the proper documentation. Guidelines in the Metasploit documentation need to be followed. For example, module files should be located in an adequate sub-directory under `modules/` directory. The two modules you submitted here are located at the root of Metasploit, which is not allowed.

**Comment by smcintyre-r7** [comment]
I'm going to close this out in favor of maintaining the existing modules. It's unclear if there's an issue with them, but if there is we should look at replication steps and get a fix in place.
